### PR TITLE
Bug 1216560 - Update change_wallpaper.py and change_layout.py in order to support the new homescreen

### PR DESCRIPTION
**Pull request by nojunpark**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9872cf7ec4c038b666f9f08fc59c4140424edcae

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py [hunk 1]
```diff
@@ -14,18 +14,18 @@
 class Homescreen(Base):
 
     name = 'Default Home Screen'
-    manifest_url = '{}verticalhome{}/manifest.webapp'.format(Base.DEFAULT_PROTOCOL,Base.DEFAULT_APP_HOSTNAME)
```
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: Are you using an older tree, I removed this already in my last pull request regarding this, I thought.

Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: oh, probably, tag me ones that I should remove.  i rebased it this morning.  I'll rebase again but I wonder whether it'll put back old code back into gaia.

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py [hunk 2]
```diff
@@ -135,15 +123,7 @@ def tap_collection(self, collection_name):
...
-        return self.marionette.execute_script("""
-        var gridItems = window.wrappedJSObject.app.grid.getItems();
-        var appElements = [];
-        for(var i=0; i<gridItems.length; i++){
-        // it must have an app to be a
-        if(gridItems[i].app) appElements.push(gridItems[i].element);
-        }
-        return appElements;
-        """)
+        return self.marionette.find_elements(*self._icons_locator)
```
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: Does this work without having to switch_to_shadow_root() ?

Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: oh this part, I just pasted it based on Johan's suggestion to avoid conflict with his PR.  I actually don't use this method.

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py [hunk 3]
```diff
@@ -14,18 +14,18 @@
...
     name = 'Default Home Screen'
-    manifest_url = '{}verticalhome{}/manifest.webapp'.format(Base.DEFAULT_PROTOCOL,Base.DEFAULT_APP_HOSTNAME)
 
     _homescreen_icon_locator = (By.CSS_SELECTOR, 'gaia-grid .icon')
     _homescreen_all_icons_locator = (By.CSS_SELECTOR, 'gaia-grid .icon:not(.placeholder)')
     _homescreen_collection_icon = (By.CSS_SELECTOR, 'gaia-grid .collection')
     _edit_mode_locator = (By.CSS_SELECTOR, 'body.edit-mode')
     _search_bar_icon_locator = (By.ID, 'search-input')
-    _landing_page_locator = (By.ID, 'icons')
+    _landing_page_locator = (By.ID, 'apps')
```
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: Is this change necessary? It looks like this locator isn't used anymore?

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py [hunk 4]
```diff
@@ -14,18 +14,18 @@
...
     _bookmark_icons_locator = (By.CSS_SELECTOR, 'gaia-grid .bookmark')
     _divider_locator = (By.CSS_SELECTOR, 'section.divider')
     _divider_separator_locator = (By.CSS_SELECTOR, 'section.divider .separator > span')
     _exit_edit_mode_locator = (By.ID, 'exit-edit-mode')
+    _icons_locator = (By.TAG_NAME, 'gaia-app-icon')
```
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: Can you add a space before this new locator? I suspect the older locators can be removed for a large part, at some point.

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py (hunk 3 above)
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: you're right, i should remove it, i didn't delete it after using it for the method that didn't work out.

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py [hunk 5]
```diff
@@ -35,8 +35,6 @@ def tap_search_bar(self):
         search_bar.tap()
 
         # TODO These lines are a workaround for bug 1020974
-        import time
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: FYI, There will be a conflict with Martijn's patch on this. He moved that function in System

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py [hunk 6]
```diff
@@ -172,9 +152,23 @@ def bookmark(self, bookmark_title):
...
+
+        _icon_container_text = 'div.gaia-container-child:nth-child(%s)'
+        change_row = False
+        #first element is always on the first column
+        _first_icon_locator = (By.CSS_SELECTOR, _icon_container_text % str(1))
+        base_x_axis = self.marionette.find_element(*_first_icon_locator).rect['x']
+
+        # loop until the row is changed (i.e., x value is same as the first icon)
+        column = 1
+        while not change_row:
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: Nit: add `is` or `has` for a boolean variable. For example: `has_row_changed` or `is_on_new_row`

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py [hunk 7]
```diff
@@ -172,9 +152,23 @@ def bookmark(self, bookmark_title):
...
-        Wait(self.marionette).until(lambda m: element.get_attribute('cols') is not None)
-        return int(element.get_attribute('cols'))
+
+        _icon_container_text = 'div.gaia-container-child:nth-child(%s)'
+        change_row = False
+        #first element is always on the first column
+        _first_icon_locator = (By.CSS_SELECTOR, _icon_container_text % str(1))
+        base_x_axis = self.marionette.find_element(*_first_icon_locator).rect['x']
+
+        # loop until the row is changed (i.e., x value is same as the first icon)
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: I think this comment is redundant with the line:
``` py
while not has_row_changed:
```
I'd remove it

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py [hunk 8]
```diff
@@ -172,9 +152,23 @@ def bookmark(self, bookmark_title):
 
     @property
     def number_of_columns(self):
-        element = self.marionette.find_element(*self._landing_page_locator)
-        Wait(self.marionette).until(lambda m: element.get_attribute('cols') is not None)
-        return int(element.get_attribute('cols'))
+
+        _icon_container_text = 'div.gaia-container-child:nth-child(%s)'
+        change_row = False
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: Nit: I'd move the definition of this variable closer to `column = 1`

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/homescreen/test_homescreen_column_layout.py [hunk 9]
```diff
@@ -30,7 +30,9 @@ def test_homescreen_column_layout(self):
         settings.launch()
 
         homescreen_settings = settings.open_homescreen()
-        homescreen_settings.select_icon_layout('Four Columns')
+        homescreen_settings.select_change_icon_layout()
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/homescreen/test_homescreen_column_layout.py]: In this test, I'd use only one function to `select` and `pick`. It'll make the test a bit easier to read I don't see a case (in our current test suite) where we tap on the button and cancel it. So we'd have:
```py
homescreen_settings.change_icon_layout('Four Columns')
```
```py
def change_icon_layout(self, layout_text):
    self._select_change_icon_layout()
    self._pick_icon_layout(text)
```
`_select_change_icon_layout()` and `_pick_icon_layout()` are protected method, as we don't use them outside anymore

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/homescreen/test_homescreen_change_wallpaper.py [hunk 10]
```diff
@@ -18,13 +18,14 @@ def test_homescreen_change_wallpaper_from_gallery(self):
...
-
         default_wallpaper_settings = self.data_layer.get_setting('wallpaper.image')
-        contextmenu = homescreen.open_context_menu()
-        activities = contextmenu.tap_change_wallpaper()
+
+        # open settings app and choose to change wallpaper, instead of opening context menu
+        settings = Settings(self.marionette)
+        settings.launch()
+        homescreen_page = settings.open_homescreen()
+        activities = homescreen_page.pick_wallpaper()
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/homescreen/test_homescreen_change_wallpaper.py]: I just realized we have the same test at `gaia/tests/python/gaia-ui-tests/gaiatest/tests/functional/settings/test_settings_wallpaper.py`. We can remove this test, then

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py [hunk 11]
```diff
@@ -71,16 +69,6 @@ def activate_edit_mode(self):
         Wait(self.marionette).until(expected.element_present(
             *self._edit_mode_locator))
 
-    def open_context_menu(self):
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: test_homescreen_change_wallpaper.py will also fail in the graphics tests.

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py (hunk 5 above)
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: I'll wait until martijn's patch is merged and rebase.

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py (hunk 11 above)
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: True, will fix it as well.

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py (hunk 7 above)
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: done

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py (hunk 6 above)
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py]: done

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/homescreen/test_homescreen_change_wallpaper.py (hunk 10 above)
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/homescreen/test_homescreen_change_wallpaper.py]: hmm, that script compares system.wallpaper_properties value, while this one compares self.data_layer.get_setting('wallpaper.image').  are they equivalent?  If so I'll remove the script

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/homescreen/test_homescreen_column_layout.py (hunk 9 above)
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/homescreen/test_homescreen_column_layout.py]: Ah, the reason I broke up the test like this is that if I bundle them together, I can't take screenshots in between.  What I'll do is create a method as you mentioned above, but make its components also available to outside as well.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5884f042650ed572bb2454965524d92f7cee5166

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=0467b498bf527e60a0cbbcf6cf9f648f6e689c61

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=48011dd192358b0ba33b20f4958f2f0514ded49f

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4c52b0aed35b02f00c17e882975c3ca7d26bf71c

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e61dd9bb4157fb5e1fd3e0e9d7dfcb51b4d6dea5
